### PR TITLE
Port 4 "streams contrib" stages

### PR DESCRIPTION
**Pull request by alex-bogomaz**
Hello, please review my attempt to port 4 following "streams contrib" stages:
- AccumulateWhileUnchanged
- LastElement
- PartitionWith
- Sample

This is for issue #3234

### CODE REVIEW

**Comment by marcpiechura** [comment]
@alex-bogomaz thanks a lot for your contribution! 

Could you approve your public api changes as described here http://getakka.net/community/public-api-changes.html ? 
I’ll review the rest of the code in the next few days.

File: src/core/Akka.Streams/Dsl/AccumulateWhileUnchanged.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                SetHandler(accumulateWhileUnchanged.In, onPush: () =>
+                {
+                    var nextElement = Grab(accumulateWhileUnchanged.In);
+                    var nextState = accumulateWhileUnchanged._propertyExtractor(nextElement);
+
+                    if (!_currentState.HasValue)
+                        _currentState = new Option<TProperty>(nextState);
+
+                    if (_currentState.Equals(new Option<TProperty>(nextState)))
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Dsl/AccumulateWhileUnchanged.cs]: Better to check _currentState.Value with nextState

File: src/core/Akka.Streams/Dsl/LastElement.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                    var element = Grab(lastElement.In);
+                    currentElement = new Option<T>(element);
+                    Push(lastElement.Out, element);
+                }, onUpstreamFinish: () =>
+                {
+                    completion.SetResult(currentElement);
+                    CompleteStage();
+                }, onUpstreamFailure: ex =>
+                {
+                    completion.SetResult(currentElement);
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Dsl/LastElement.cs]: Shouldn’t that be SetException?

File: src/core/Akka.Streams/Dsl/LastElement.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                    CompleteStage();
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Dsl/LastElement.cs]: And this FailStage?

File: src/core/Akka.Streams/Dsl/LastElement.cs (hunk 2 above)
Comment (alex-bogomaz) [reply src/core/Akka.Streams/Dsl/LastElement.cs]: in Scala version they are not failing stage but completing it upon exception:
https://github.com/akka/akka-stream-contrib/blob/ea7a0ce0dd320a2d3c443fa6097a517b20b31c13/contrib/src/main/scala/akka/stream/contrib/LastElement.scala#L63

this is also captured in the unit test ```materialize to the last element emitted by a source before it failed```:
https://github.com/akka/akka-stream-contrib/blob/ab0817a660906316e282d0c7c762cb1d18a05c5d/contrib/src/test/scala/akka/stream/contrib/LastElementSpec.scala#L41

Comment (marcpiechura) [reply src/core/Akka.Streams/Dsl/LastElement.cs]: @alex-bogomaz alright, thx for the clarification, will merge once build is green 👍

**Pull request merged**